craft_providers/multipass/multipass.py

```diff
@@ -0,0 +1,382 @@
+import json
+import logging
+import pathlib
+import shlex
+import subprocess
+from typing import Any, Dict, List, Optional
+
+from craft_providers import errors
+
+from .errors import MultipassError
+
+logger = logging.getLogger(__name__)
+
+
+class Multipass:
+    """Wrapper for multipass command.
+
+    :param multipass_path: Path to multipass command to use.
+    """
+
+    def __init__(
+        self, *, multipass_path: pathlib.Path = pathlib.Path("multipass")
+    ) -> None:
+        self.multipass_path = multipass_path
+
+    def _run(  # pylint: disable=redefined-builtin
+        self,
+        command: List[str],
+        *,
+        check=True,
+        **kwargs,
+    ) -> subprocess.CompletedProcess:
+        """Execute command in instance_name, allowing output to console."""
+        command = [str(self.multipass_path), *command]
+
+        logger.debug("Executing on host: %s", shlex.join(command))
+        return subprocess.run(command, check=check, **kwargs)
+
+    def delete(self, *, instance_name: str, purge=True) -> None:
+        """Passthrough for running multipass delete.
+
+        :param instance_name: The name of the instance_name to delete.
+        :param purge: Flag to purge the instance_name's image after deleting.
+
+        :raises MultipassError: on error.
+
+        """
+        command = ["delete", instance_name]
+        if purge:
+            command.append("--purge")
+
+        try:
+            self._run(command, capture_output=True, check=True)
+        except subprocess.CalledProcessError as error:
+            raise MultipassError(
+                brief=f"Failed to delete VM {instance_name!r}.",
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def exec(
+        self,
+        *,
+        command: List[str],
+        instance_name: str,
+        runner=subprocess.run,
+        **kwargs,
+    ):
+        """Execute command in instance_name with specified runner.
+
+        :param command: Command to execute in the instance.
+        :param instance_name: Name of instance to execute in.
+
+        :returns: Runner's instance.
+        """
+        final_cmd = [str(self.multipass_path), "exec", instance_name, "--", *command]
+
+        quoted_final_cmd = shlex.join(final_cmd)
+        logger.debug("Executing on host: %s", quoted_final_cmd)
+
+        return runner(final_cmd, **kwargs)  # pylint: disable=subprocess-run-check
+
+    def info(self, *, instance_name: str) -> Optional[Dict[str, Any]]:
+        """Get information/state for instance.
+
+        :returns: Parsed json data from info command.
+
+        :raises MultipassError: On error.
+        """
+        command = ["info", instance_name, "--format", "json"]
+
+        try:
+            proc = self._run(command, capture_output=True, check=True, text=True)
+        except subprocess.CalledProcessError as error:
+            if "does not exist" in error.stderr:
+                return None
```

Comment:
Why are we hiding this error?

Reply:
It was the approach used to determine if the VM exists...

However, on second thought, I will rewrite that logic to rely on list().

Dropped.